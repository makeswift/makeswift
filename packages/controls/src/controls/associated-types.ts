export type AssociatedTypes<T> = T extends {
  __associatedTypes(types: infer Types): void
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]

export type ControlType<D> = AssociatedType<D, 'ControlType'>
export type ConfigType<D> = AssociatedType<D, 'Config'>
export type DataType<D> = AssociatedType<D, 'DataType'>
export type ValueType<D> = AssociatedType<D, 'ValueType'>
export type ResolvedValueType<D> = AssociatedType<D, 'ResolvedValueType'>
export type InstanceType<D> = AssociatedType<D, 'InstanceType'>
