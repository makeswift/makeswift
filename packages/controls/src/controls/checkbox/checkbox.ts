import { match, P } from 'ts-pattern'
import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey } from '../../common'
import { type CopyContext } from '../../context'
import { type DeserializedRecord } from '../../serialization'

import {
  ControlDefinition,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'
import { ControlDefinitionVisitor } from '../visitor'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'checkbox::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::checkbox' as const

  static get schema() {
    const version = z.literal(1).optional()
    const versionedData = z.object({
      [ControlDataTypeKey]: z.literal(this.v1DataType),
      value: z.boolean(),
    })

    const schemas = <V, D>(value: SchemaType<V>, data: SchemaType<D>) => {
      const type = z.literal(this.type)

      const config = z.object({
        label: z.string().optional(),
        description: z.string().optional(),
        defaultValue: value,
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedValue: value,
        data,
        versionedData,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(
        z.boolean().optional(),
        z.union([z.boolean(), versionedData, z.undefined()]),
      ),
      strict: schemas(z.boolean(), z.union([z.boolean(), versionedData])),
    }
  }

  static deserialize(data: DeserializedRecord): CheckboxDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Checkbox: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new CheckboxDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ value }) => value)
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match({ version: this.version, value })
      .with({ version: 1, value: P.boolean }, ({ value }) => ({
        ...Definition.dataSignature.v1,
        value,
      }))
      .with({ version: 1, value: undefined }, () => undefined)
      .with({ version: undefined }, ({ value }) => value)
      .otherwise(() => value)
  }

  copyData(
    data: DataType<C> | undefined,
    _context: CopyContext,
  ): DataType<C> | undefined {
    return data
  }

  resolveValue(
    data: DataType<C> | undefined,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    return {
      name: Definition.type,
      readStable: () => this.fromData(data) ?? this.config.defaultValue,
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  accept<R>(visitor: ControlDefinitionVisitor<R>, ...args: unknown[]): R {
    return visitor.visitCheckbox(this, ...args)
  }
}

export class CheckboxDefinition<
  C extends Config = Config,
> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Checkbox<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): CheckboxDefinition<NormedConfig<D>> {
  return new CheckboxDefinition((config ?? {}) as NormedConfig<D>, 1)
}
