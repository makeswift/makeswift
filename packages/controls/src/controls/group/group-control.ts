import { ControlDefinition } from '../definition'
import {
  ControlInstance,
  type ControlMessage,
  type SendMessage,
} from '../instance'

import { type GroupDefinition } from './group'

type Message = {
  type: typeof GroupControl.CHILD_CONTROL_MESSAGE
  payload: { message: ControlMessage; key: string }
}

export class GroupControl<
  Def extends GroupDefinition = GroupDefinition,
> extends ControlInstance<Message> {
  static readonly CHILD_CONTROL_MESSAGE =
    'makeswift::controls::group::message::child-control-message'

  private readonly childControls: Map<string, ControlInstance> = new Map()

  constructor(
    private readonly definition: Def,
    sendMessage: SendMessage<Message>,
  ) {
    super(sendMessage)
    this.childControls = new Map(
      Object.entries(this.definition.propDefs).map(([key, def]) => [
        key,
        this.createChildControl(def, key),
      ]),
    )
  }

  recv = (message: Message) => {
    switch (message.type) {
      case GroupControl.CHILD_CONTROL_MESSAGE: {
        this.child(message.payload.key)?.recv(message.payload.message)
      }
    }
  }

  child = (key: string) => this.childControls.get(key)

  createChildControl = (def: ControlDefinition, key: string) => {
    return def.createInstance((message) =>
      this.sendMessage({
        type: GroupControl.CHILD_CONTROL_MESSAGE,
        payload: { message, key },
      }),
    )
  }
}
