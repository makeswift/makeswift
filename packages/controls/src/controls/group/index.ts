export * from './group'
export * from './group-control'
