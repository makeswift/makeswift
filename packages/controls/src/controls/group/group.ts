import { z } from 'zod'

import { mapValues } from '../../lib/functional'
import { StableValue } from '../../lib/stable-value'
import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey, type Data } from '../../common'
import {
  type CopyContext,
  type MergeTranslatableDataContext,
} from '../../context'
import { type IntrospectionTarget } from '../../introspection'
import { type ResourceResolver } from '../../resources/resolver'
import {
  SerializationSchema,
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'
import { type Stylesheet } from '../../stylesheet'

import {
  type DataType as DataType_,
  type ResolvedValueType as ResolvedValueType_,
  type ValueType as ValueType_,
} from '../associated-types'
import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { type SendMessage } from '../instance'
import { ShapeDefinition } from '../shape/v1'
import { ShapeV2Definition } from '../shape/v2'

import { GroupControl } from './group-control'

type ItemDefinition = ControlDefinition<string, unknown, any, any, any>
type KeyDefinitions = Record<string, ItemDefinition>

type Config<Defs extends KeyDefinitions = KeyDefinitions> = {
  readonly label?: string
  readonly preferredLayout?:
    | typeof Definition.Layout.Inline
    | typeof Definition.Layout.Popover
  readonly props: Defs
}

type ShapeData<C extends Config> = DataType_<
  ShapeDefinition<{ type: C['props'] }>
>

type PropsData<C extends Config> = {
  [K in keyof C['props']]: DataType_<C['props'][K]>
}

type VersionedData<C extends Config> = {
  [ControlDataTypeKey]:
    | typeof Definition.v1DataType
    | typeof ShapeV2Definition.v1DataType
  value: PropsData<C>
}

type DataType<C extends Config> = ShapeData<C> | VersionedData<C>

type ValueType<C extends Config> = {
  [K in keyof C['props']]: ValueType_<C['props'][K]>
}

type ResolvedValueType<C extends Config> = {
  [K in keyof C['props']]: ResolvedValueType_<C['props'][K]>
}

type InstanceType<C extends Config> = GroupControl<Definition<C>>

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>,
  InstanceType<C>
> {
  static readonly type = 'makeswift::controls::group' as const
  static readonly v1DataType = 'group::v1' as const

  static readonly Layout = {
    Inline: `${this.type}::layout::inline`,
    Popover: `${this.type}::layout::popover`,
  } as const

  static deserialize(
    data: DeserializedRecord,
    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
  ): GroupDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Group: expected '${Definition.type}', got '${data.type}'`,
      )

    const {
      config: { props, ...config },
    } = z
      .object({
        type: z.literal(Definition.type),
        config: z.object({
          label: z.string().optional(),
          preferredLayout: z
            .union([
              z.literal(Definition.Layout.Inline),
              z.literal(Definition.Layout.Popover),
            ])
            .optional(),
          props: z.record(z.string(), SerializationSchema.deserializedRecord),
        }),
      })
      .parse(data)

    const deserializedProps = mapValues(props, (itemDef) => {
      return deserializeCallback(itemDef)
    })

    return Group({
      props: deserializedProps,
      ...config,
    })
  }

  constructor(
    config: C,
    readonly dataType:
      | typeof Definition.v1DataType
      | typeof ShapeV2Definition.v1DataType,
  ) {
    super(config)
  }

  private get dataSignature() {
    return { [ControlDataTypeKey]: this.dataType }
  }

  get propDefs() {
    return this.config.props
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    const version = z.literal(1)
    const type = z.literal(Definition.type)

    const keys = mapValues(this.propDefs, (def) => def.schema)

    const propsData = z.object(
      mapValues(keys, (key) => key.data),
    ) as SchemaType<PropsData<C>>

    const versionedData = z.object({
      [ControlDataTypeKey]: z.union([
        z.literal(Definition.v1DataType),
        z.literal(ShapeV2Definition.v1DataType),
      ]),
      value: propsData,
    }) as SchemaType<VersionedData<C>>

    const data = z.union([propsData, versionedData])

    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
      ValueType<C>
    >

    const resolvedValue = z.object(
      mapValues(keys, (key) => key.resolvedValue),
    ) as SchemaType<ResolvedValueType<C>>

    const config = z.object({
      label: z.string().optional(),
      preferredLayout: z.union([
        z.literal(Definition.Layout.Inline),
        z.literal(Definition.Layout.Popover),
      ]),
      props: z.object(mapValues(keys, (key) => key.definition)),
    })

    const definition = z.object({
      type,
      config,
    })

    return { type, data, value, resolvedValue, definition, version }
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    if (data == null) return undefined

    const propsData = Definition.propsData(data)

    return mapValues(this.propDefs, (def, key) =>
      def.fromData(propsData[key]),
    ) as ValueType<C>
  }

  propsToData(propsData: PropsData<C>): DataType<C> {
    return {
      ...this.dataSignature,
      value: propsData,
    }
  }

  toData(value: ValueType<C>): DataType<C> {
    return this.propsToData(
      mapValues(this.propDefs, (def, key) => def.toData(value[key])),
    )
  }

  copyData(
    data: DataType<C> | undefined,
    context: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return undefined

    const propsData = Definition.propsData(data)

    return this.propsToData(
      mapValues(this.propDefs, (def, key) =>
        def.copyData(propsData[key], context),
      ),
    )
  }

  getTranslatableData(data: DataType<C>): Data {
    if (data == null) return null

    const propsData = Definition.propsData(data)

    return mapValues(this.propDefs, (def, key) =>
      def.getTranslatableData(propsData[key]),
    )
  }

  mergeTranslatedData(
    data: DataType<C>,
    translatedData: Record<string, DataType<C>>,
    context: MergeTranslatableDataContext,
  ): Data {
    if (translatedData == null) return data

    const propsData = Definition.propsData(data)

    return mapValues(this.propDefs, (def, key) =>
      def.mergeTranslatedData(propsData[key], translatedData[key], context),
    )
  }

  resolveValue(
    data: DataType<C> | undefined,
    resolver: ResourceResolver,
    stylesheet: Stylesheet,
    control?: InstanceType<C>,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    const propsData = data != null ? Definition.propsData(data) : undefined

    const keyValues = mapValues(this.propDefs, (def, key) =>
      def.resolveValue(
        propsData?.[key],
        resolver,
        stylesheet.child(key),
        control?.child(key),
      ),
    )

    const stableValue = StableValue({
      name: Definition.type,
      read: () =>
        mapValues(keyValues, (v) => v.readStable()) as ResolvedValueType<C>,
      deps: Object.values(keyValues),
    })

    return {
      ...stableValue,
      triggerResolve: async (currentValue?: ResolvedValueType<C>) => {
        await Promise.all(
          Object.entries(keyValues).map(([key, v]) =>
            v?.triggerResolve(currentValue?.[key]),
          ),
        )
      },
    }
  }

  createInstance(sendMessage: SendMessage): InstanceType<C> {
    return new GroupControl(this, sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  introspect<R>(data: DataType<C> | undefined, target: IntrospectionTarget<R>) {
    if (data == null) return []

    const propsData = Definition.propsData(data)

    return Object.entries(this.propDefs).flatMap(
      ([key, def]) => def.introspect(propsData[key], target) ?? [],
    )
  }

  static propsData<C extends Config>(data: DataType<C>): PropsData<C> {
    if (Definition.isVersionedData(data)) {
      return data.value
    }

    return data
  }

  static propDataPath<C extends Config>(
    data: DataType<C>,
    key: string,
  ): string[] {
    return Definition.isVersionedData(data) ? ['value', key] : [key]
  }

  private static isVersionedData<C extends Config>(
    data: DataType<C>,
  ): data is VersionedData<C> {
    return (
      data != null &&
      ControlDataTypeKey in data &&
      (data[ControlDataTypeKey] === Definition.v1DataType ||
        data[ControlDataTypeKey] === ShapeV2Definition.v1DataType)
    )
  }
}

export class GroupDefinition<C extends Config = Config> extends Definition<C> {}

export function Group<D extends KeyDefinitions>(
  config: Config<D>,
): GroupDefinition<Config<D>> {
  return new GroupDefinition<Config<D>>(
    config?.preferredLayout == null
      ? { preferredLayout: Definition.Layout.Inline, ...config }
      : config,
    Definition.v1DataType,
  )
}

Group.Layout = Definition.Layout
