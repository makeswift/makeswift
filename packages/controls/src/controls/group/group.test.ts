import { ControlDataTypeKey } from '../../common'
import { createReplacementContext } from '../../context'
import { Targets } from '../../introspection'
import { deserializeRecord, type DeserializedRecord } from '../../serialization'

import { type DataType, type ValueType } from '../associated-types'
import { Checkbox, CheckboxDefinition } from '../checkbox'
import { Color, ColorDefinition } from '../color'
import { Combobox } from '../combobox'
import { ControlDefinition } from '../definition'
import { Image } from '../image'
import { GenericLink as Link } from '../link'
import { List } from '../list'
import { Number } from '../number'
import { TextArea } from '../text-area'
import { TextInput } from '../text-input'

import { Group, GroupDefinition } from './group'

function deserializer(record: DeserializedRecord): ControlDefinition {
  switch (record.type) {
    case GroupDefinition.type:
      return GroupDefinition.deserialize(record, deserializer)

    case ColorDefinition.type:
      return ColorDefinition.deserialize(record)

    case CheckboxDefinition.type:
      return CheckboxDefinition.deserialize(record)

    default:
      throw new Error('Invalid control type')
  }
}

describe('Group', () => {
  describe('constructor', () => {
    test('heterogenous controls', () => {
      const group = Group({
        props: {
          val: Color({ defaultValue: 'red' }),
          num: Checkbox(),
        },
      })
      expect(group).toMatchSnapshot()
    })

    test('disallows extraneous properties', () => {
      Group({
        props: {
          toggle: Checkbox(),
        },
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: GroupDefinition) {}
    assignTest(
      Group({
        props: {},
        label: 'Group',
      }),
    )

    assignTest(
      Group({
        props: {},
      }),
    )

    assignTest(
      Group({
        props: {
          toggle: Checkbox(),
          color: Color({ defaultValue: 'red' }),
        },
      }),
    )

    assignTest(
      Group({
        props: {
          combo: Combobox({
            getOptions: () => [{ id: '0', value: 'a', label: 'Alpha' }],
          }),
          link: Link({ label: 'Link' }),
        },
      }),
    )
  })

  describe('serialization', () => {
    test('serialize/deserialize layout of controls', () => {
      const group = Group({
        props: {},
        preferredLayout: 'makeswift::controls::group::layout::popover',
      })

      const [serialized, _] = group.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = GroupDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(group)
    })

    test('serialize/deserialize label of controls', () => {
      const group = Group({
        props: {},
        label: 'Group',
      })

      const [serialized, _] = group.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = GroupDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(group)
    })

    test('serialize/deserialize group of controls', () => {
      const group = Group({
        props: {
          color: Color({ defaultValue: 'red' }),
          checkbox: Checkbox(),
        },
      })

      const [serialized, _] = group.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = GroupDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(group)
    })

    test('serialize/deserialize composable group of controls', () => {
      const group = Group({
        props: {
          group: Group({
            props: {
              color: Color({ defaultValue: 'red' }),
              checkbox: Checkbox(),
            },
          }),
        },
      })

      const [serialized, _] = group.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = GroupDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(group)
    })
  })

  describe('introspection', () => {
    test('Group data is correctly delegated to subcontrols', () => {
      const group = Group({
        props: {
          color: Color({ defaultValue: 'red' }),
          link: Link(),
          image: Image(),
        },
      })

      const groupData = {
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
        link: {
          type: 'OPEN_PAGE' as const,
          payload: {
            pageId: 'page-id',
            openInNewTab: false,
          },
        },
        image: 'file-id',
      }

      const swatchIds = group.introspect(groupData, Targets.Swatch)
      expect(swatchIds).toEqual(['swatch-id'])

      const pageIds = group.introspect(groupData, Targets.Page)
      expect(pageIds).toEqual(['page-id'])

      const fileIds = group.introspect(groupData, Targets.File)
      expect(fileIds).toEqual(['file-id'])
    })

    test('Group versioned data is correctly delegated to subcontrols', () => {
      const group = Group({
        props: {
          color: Color({ defaultValue: 'red' }),
          link: Link(),
          image: Image(),
        },
      })

      const groupData = {
        [ControlDataTypeKey]: 'group::v1' as const,
        value: {
          color: {
            swatchId: 'swatch-id',
            alpha: 1,
          },
          link: {
            type: 'OPEN_PAGE' as const,
            payload: {
              pageId: 'page-id',
              openInNewTab: false,
            },
          },
          image: 'file-id',
        },
      }

      const swatchIds = group.introspect(groupData, Targets.Swatch)
      expect(swatchIds).toEqual(['swatch-id'])

      const pageIds = group.introspect(groupData, Targets.Page)
      expect(pageIds).toEqual(['page-id'])

      const fileIds = group.introspect(groupData, Targets.File)
      expect(fileIds).toEqual(['file-id'])
    })
  })

  describe('getTranslatableData', () => {
    test('Group data is correctly delegated to subcontrols', () => {
      const group = Group({
        props: {
          textArea: TextArea(),
          textInput: TextInput(),
          color: Color({ defaultValue: 'red' }),
        },
      })

      const groupData = {
        textArea: 'alpha',
        textInput: 'beta',
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
      }

      const translatableData = group.getTranslatableData(groupData)
      expect(translatableData).toEqual({
        textArea: 'alpha',
        textInput: 'beta',
        color: null,
      })
    })

    test('Group versioned data is correctly delegated to subcontrols', () => {
      const group = Group({
        props: {
          textArea: TextArea(),
          textInput: TextInput(),
          color: Color({ defaultValue: 'red' }),
        },
      })

      const groupData = {
        [ControlDataTypeKey]: 'group::v1' as const,
        value: {
          textArea: 'alpha',
          textInput: 'beta',
          color: {
            swatchId: 'swatch-id',
            alpha: 1,
          },
        },
      }

      const translatableData = group.getTranslatableData(groupData)
      expect(translatableData).toEqual({
        textArea: 'alpha',
        textInput: 'beta',
        color: null,
      })
    })
  })

  describe('Group data gracefully handles missing/extra props', () => {
    const group = Group({
      props: {
        color: Color({ defaultValue: 'red' }),
        list: List({ type: Number() }),
        link: Link(),
        text: TextArea(),
      },
    })

    const data = {
      // missing props + extra prop that should be ignored
      color: { swatchId: '[swatch-id]', alpha: 1 },
      text: 'Hello world',
      extraProp: 'extra',
    } as any as DataType<typeof group>

    const value = {
      color: { swatchId: '[swatch-id]', alpha: 1 },
      list: [17],
      link: {
        type: 'OPEN_PAGE',
        payload: { pageId: '[page-id]', openInNewTab: false },
      },
      text: 'Hello world',
      // extra prop only, `toData` doesn't handle missing props
      extraProp: 'extra',
    } as ValueType<typeof group>

    test('fromData', () => expect(group.fromData(data)).toMatchSnapshot())
    test('toData', () => expect(group.toData(value)).toMatchSnapshot())
    test('copyData', () =>
      expect(
        group.copyData(data, {
          replacementContext: createReplacementContext({}),
          copyElement: (node) => node,
        }),
      ).toMatchSnapshot())

    test('getTranslatableData', () =>
      expect(group.getTranslatableData(data)).toMatchSnapshot())

    test('mergeTranslatedData', () =>
      expect(
        group.mergeTranslatedData(
          data,
          {},
          {
            translatedData: {},
            mergeTranslatedData: (node) => node,
          },
        ),
      ).toMatchSnapshot())

    test('introspect', () =>
      expect(group.introspect(data, Targets.ChildrenElement)).toStrictEqual([]))
  })

  describe('Group versioned data is gracefully handles missing/extra props', () => {
    const group = Group({
      props: {
        color: Color({ defaultValue: 'red' }),
        list: List({ type: Number() }),
        link: Link(),
        text: TextArea(),
      },
    })

    const data = {
      [ControlDataTypeKey]: 'group::v1' as const,
      // missing props + extra prop that should be ignored
      value: {
        color: { swatchId: '[swatch-id]', alpha: 1 },
        text: 'Hello world',
        extraProp: 'extra',
      },
    } as any as DataType<typeof group>

    const value = {
      color: { swatchId: '[swatch-id]', alpha: 1 },
      list: [17],
      link: {
        type: 'OPEN_PAGE',
        payload: { pageId: '[page-id]', openInNewTab: false },
      },
      text: 'Hello world',
      // extra prop only, `toData` doesn't handle missing props
      extraProp: 'extra',
    } as ValueType<typeof group>

    test('fromData', () => expect(group.fromData(data)).toMatchSnapshot())
    test('toData', () => expect(group.toData(value)).toMatchSnapshot())
    test('copyData', () =>
      expect(
        group.copyData(data, {
          replacementContext: createReplacementContext({}),
          copyElement: (node) => node,
        }),
      ).toMatchSnapshot())

    test('getTranslatableData', () =>
      expect(group.getTranslatableData(data)).toMatchSnapshot())

    test('mergeTranslatedData', () =>
      expect(
        group.mergeTranslatedData(
          data,
          {},
          {
            translatedData: {},
            mergeTranslatedData: (node) => node,
          },
        ),
      ).toMatchSnapshot())

    test('introspect', () =>
      expect(group.introspect(data, Targets.ChildrenElement)).toStrictEqual([]))
  })
})
