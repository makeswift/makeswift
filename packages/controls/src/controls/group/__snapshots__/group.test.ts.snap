// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group Group data gracefully handles missing/extra props copyData 1`] = `
{
  "@@makeswift/type": "group::v1",
  "value": {
    "color": {
      "alpha": 1,
      "swatchId": "[swatch-id]",
    },
    "link": undefined,
    "list": undefined,
    "text": "Hello world",
  },
}
`;

exports[`Group Group data gracefully handles missing/extra props fromData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Group Group data gracefully handles missing/extra props getTranslatableData 1`] = `
{
  "color": null,
  "link": null,
  "list": null,
  "text": "Hello world",
}
`;

exports[`Group Group data gracefully handles missing/extra props mergeTranslatedData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Group Group data gracefully handles missing/extra props toData 1`] = `
{
  "@@makeswift/type": "group::v1",
  "value": {
    "color": {
      "@@makeswift/type": "color::v1",
      "alpha": 1,
      "swatchId": "[swatch-id]",
    },
    "link": {
      "payload": {
        "openInNewTab": false,
        "pageId": "[page-id]",
      },
      "type": "OPEN_PAGE",
    },
    "list": [
      {
        "id": "xxxxxxxx-xxxx-xxxx-xxxx-100000000000",
        "type": "makeswift::controls::number",
        "value": {
          "@@makeswift/type": "number::v1",
          "value": 17,
        },
      },
    ],
    "text": {
      "@@makeswift/type": "text-area::v1",
      "value": "Hello world",
    },
  },
}
`;

exports[`Group Group versioned data is gracefully handles missing/extra props copyData 1`] = `
{
  "@@makeswift/type": "group::v1",
  "value": {
    "color": {
      "alpha": 1,
      "swatchId": "[swatch-id]",
    },
    "link": undefined,
    "list": undefined,
    "text": "Hello world",
  },
}
`;

exports[`Group Group versioned data is gracefully handles missing/extra props fromData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Group Group versioned data is gracefully handles missing/extra props getTranslatableData 1`] = `
{
  "color": null,
  "link": null,
  "list": null,
  "text": "Hello world",
}
`;

exports[`Group Group versioned data is gracefully handles missing/extra props mergeTranslatedData 1`] = `
{
  "color": {
    "alpha": 1,
    "swatchId": "[swatch-id]",
  },
  "link": undefined,
  "list": undefined,
  "text": "Hello world",
}
`;

exports[`Group Group versioned data is gracefully handles missing/extra props toData 1`] = `
{
  "@@makeswift/type": "group::v1",
  "value": {
    "color": {
      "@@makeswift/type": "color::v1",
      "alpha": 1,
      "swatchId": "[swatch-id]",
    },
    "link": {
      "payload": {
        "openInNewTab": false,
        "pageId": "[page-id]",
      },
      "type": "OPEN_PAGE",
    },
    "list": [
      {
        "id": "xxxxxxxx-xxxx-xxxx-xxxx-100000000001",
        "type": "makeswift::controls::number",
        "value": {
          "@@makeswift/type": "number::v1",
          "value": 17,
        },
      },
    ],
    "text": {
      "@@makeswift/type": "text-area::v1",
      "value": "Hello world",
    },
  },
}
`;

exports[`Group constructor heterogenous controls 1`] = `
GroupDefinition {
  "config": {
    "preferredLayout": "makeswift::controls::group::layout::inline",
    "props": {
      "num": CheckboxDefinition {
        "config": {},
        "version": 1,
      },
      "val": ColorDefinition {
        "config": {
          "defaultValue": "red",
        },
        "version": 1,
      },
    },
  },
  "dataType": "group::v1",
}
`;

exports[`Group serialization serialize/deserialize composable group of controls 1`] = `
{
  "config": {
    "preferredLayout": "makeswift::controls::group::layout::inline",
    "props": {
      "group": {
        "config": {
          "preferredLayout": "makeswift::controls::group::layout::inline",
          "props": {
            "checkbox": {
              "config": {},
              "type": "makeswift::controls::checkbox",
              "version": 1,
            },
            "color": {
              "config": {
                "defaultValue": "red",
              },
              "type": "makeswift::controls::color",
              "version": 1,
            },
          },
        },
        "type": "makeswift::controls::group",
      },
    },
  },
  "type": "makeswift::controls::group",
}
`;

exports[`Group serialization serialize/deserialize group of controls 1`] = `
{
  "config": {
    "preferredLayout": "makeswift::controls::group::layout::inline",
    "props": {
      "checkbox": {
        "config": {},
        "type": "makeswift::controls::checkbox",
        "version": 1,
      },
      "color": {
        "config": {
          "defaultValue": "red",
        },
        "type": "makeswift::controls::color",
        "version": 1,
      },
    },
  },
  "type": "makeswift::controls::group",
}
`;

exports[`Group serialization serialize/deserialize label of controls 1`] = `
{
  "config": {
    "label": "Group",
    "preferredLayout": "makeswift::controls::group::layout::inline",
    "props": {},
  },
  "type": "makeswift::controls::group",
}
`;

exports[`Group serialization serialize/deserialize layout of controls 1`] = `
{
  "config": {
    "preferredLayout": "makeswift::controls::group::layout::popover",
    "props": {},
  },
  "type": "makeswift::controls::group",
}
`;
